[FSDP] full_state_dict impl

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73324
* #73323

Implements `state_dict` and `load_state_dict` APIs for FSDP, with the following limitations:

1. Does not support `state_dict_device` (i.e. specifying which device params should be on) which fairscale does currently support
2. Does not yet support offload of state_dict onto CPU
3. Loads state_dict on all ranks currently. In the future we could add support for loading this on only rank 0, to avoid redundancy across ranks as usually only one rank is responsible for saving/loading the model. Along with (2) this would enable larger models to have state_dict called.

As discussed in FSDP checkpoint API proposal, `state_dict` will basically be a `full_state_dict` where full parameters are returned on all ranks. As a result this implies that the model must actually be able to fit on a single GPU.

Differential Revision: [D34433514](https://our.internmc.facebook.com/intern/diff/D34433514/)

rohan-varma committed:
[FSDP] full_state_dict impl

Implements `state_dict` and `load_state_dict` APIs for FSDP, with the following limitations:

1. Does not support `state_dict_device` (i.e. specifying which device params should be on) which fairscale does currently support
2. Does not yet support offload of state_dict onto CPU
3. Loads state_dict on all ranks currently. In the future we could add support for loading this on only rank 0, to avoid redundancy across ranks as usually only one rank is responsible for saving/loading the model. Along with (2) this would enable larger models to have state_dict called.

As discussed in FSDP checkpoint API proposal, `state_dict` will basically be a `full_state_dict` where full parameters are returned on all ranks. As a result this implies that the model must actually be able to fit on a single GPU.

Differential Revision: [D34433514](https://our.internmc.facebook.com/intern/diff/D34433514/)

[ghstack-poisoned]

fegin commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -58,14 +69,109 @@ class TestFSDPStateDict(FSDPTest):
+    def test_basic_save_and_load_full_state_dict(self, cpu_offload, fp16):
+        """
+        Tests that we can save a full_state_dict and load it into a blank model
+        with various configs such as fp17 and cpu offload and parameters
```
typo, fp16?

fegin replied:
typo, fp16?

fegin commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -58,14 +69,109 @@ class TestFSDPStateDict(FSDPTest):
+            model = model_call()
+            fsdp_state_dict = _get_state_dict(model, cpu_offload.offload_params, fp16)
+            if fp16:
+                # Verify fp17 is the type
```
ditto

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -846,7 +862,17 @@ def state_dict(self, *args, **kwargs):
             torch.cuda.synchronize()
         if self._state_dict_type == StateDictType.FULL_STATE_DICT:
-            return super().state_dict(*args, **kwargs)
+            self._lazy_init()
```
Can we move this outside of the `if` to ensure all different state_dict call `_lazy_init`?

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -776,6 +776,22 @@ def _full_post_state_dict_hook(
+        for key in state_dict.keys():
+            # Due to recursive call of summon_full_params, avoid unnecessary reclone of
+            if (
+                self.training_state == TrainingState_.SUMMON_FULL_PARAMS and
```
Curious, is there any chance that this hook will be called with `self.training_state` is not `SUMMON_FULL_PARAMS`?

rohan-varma replied:
Good point, it should never be called as far as I understand. I'll see if I can actually just change this to an assert.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -776,6 +776,22 @@ def _full_post_state_dict_hook(
+                self.training_state == TrainingState_.SUMMON_FULL_PARAMS and
+                not getattr(state_dict[key], "_has_been_cloned", False)
+            ):
+                state_dict[key] = state_dict[key].clone()
```
Is `detach()` needed here?

rohan-varma replied:
good point, looks like .clone().detach() is the typical pattern for tensor cloning.

fegin reviewed (commented):
LGTM, just a few nits.

rohan-varma committed:
Update on "[FSDP] full_state_dict impl"


Implements `state_dict` and `load_state_dict` APIs for FSDP, with the following limitations:

1. Does not support `state_dict_device` (i.e. specifying which device params should be on) which fairscale does currently support
2. Does not yet support offload of state_dict onto CPU
3. Loads state_dict on all ranks currently. In the future we could add support for loading this on only rank 0, to avoid redundancy across ranks as usually only one rank is responsible for saving/loading the model. Along with (2) this would enable larger models to have state_dict called.

As discussed in FSDP checkpoint API proposal, `state_dict` will basically be a `full_state_dict` where full parameters are returned on all ranks. As a result this implies that the model must actually be able to fit on a single GPU.

Differential Revision: [D34433514](https://our.internmc.facebook.com/intern/diff/D34433514/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -845,8 +862,18 @@ def state_dict(self, *args, **kwargs):
+            else:
+                state_dict = super().state_dict(*args, **kwargs)
+
+            # TODO: support offload to CPU in post state dict hook.
```
will add the issue

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -845,8 +862,18 @@ def state_dict(self, *args, **kwargs):
+                with self._summon_full_params(recurse=False, writeback=False):
+                    state_dict = super().state_dict(*args, **kwargs)
+            else:
+                state_dict = super().state_dict(*args, **kwargs)
```
need to add a UT where we are already in summon_full_params context

rohan-varma committed:
Update on "[FSDP] full_state_dict impl"


Implements `state_dict` and `load_state_dict` APIs for FSDP, with the following limitations:

1. Does not support `state_dict_device` (i.e. specifying which device params should be on) which fairscale does currently support
2. Does not yet support offload of state_dict onto CPU
3. Loads state_dict on all ranks currently. In the future we could add support for loading this on only rank 0, to avoid redundancy across ranks as usually only one rank is responsible for saving/loading the model. Along with (2) this would enable larger models to have state_dict called.

As discussed in FSDP checkpoint API proposal, `state_dict` will basically be a `full_state_dict` where full parameters are returned on all ranks. As a result this implies that the model must actually be able to fit on a single GPU.

Differential Revision: [D34433514](https://our.internmc.facebook.com/intern/diff/D34433514/)

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -865,12 +892,27 @@ def local_state_dict(self, *args: Any, **kwargs: Any) -> Any:
         with self.state_dict_type(StateDictType.LOCAL_STATE_DICT):
             return self.state_dict(*args, **kwargs)
 
+    def full_state_dict(self, *args: Any, **kwargs: Any) -> Any:
```
synced up offline, we can remove full_state_dict and load_full_state_dict interfaces

zhaojuanmao reviewed (commented):
looks good, just one minor comment regarding removing full_state_dict() interface as synced offline

rohan-varma committed:
Update on "[FSDP] full_state_dict impl"


Implements `state_dict` and `load_state_dict` APIs for FSDP, with the following limitations:

1. Does not support `state_dict_device` (i.e. specifying which device params should be on) which fairscale does currently support
2. Does not yet support offload of state_dict onto CPU
3. Loads state_dict on all ranks currently. In the future we could add support for loading this on only rank 0, to avoid redundancy across ranks as usually only one rank is responsible for saving/loading the model. Along with (2) this would enable larger models to have state_dict called.

As discussed in FSDP checkpoint API proposal, `state_dict` will basically be a `full_state_dict` where full parameters are returned on all ranks. As a result this implies that the model must actually be able to fit on a single GPU.

Differential Revision: [D34433514](https://our.internmc.facebook.com/intern/diff/D34433514/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] full_state_dict impl"


Implements `state_dict` and `load_state_dict` APIs for FSDP, with the following limitations:

1. Does not support `state_dict_device` (i.e. specifying which device params should be on) which fairscale does currently support
2. Does not yet support offload of state_dict onto CPU
3. Loads state_dict on all ranks currently. In the future we could add support for loading this on only rank 0, to avoid redundancy across ranks as usually only one rank is responsible for saving/loading the model. Along with (2) this would enable larger models to have state_dict called.

As discussed in FSDP checkpoint API proposal, `state_dict` will basically be a `full_state_dict` where full parameters are returned on all ranks. As a result this implies that the model must actually be able to fit on a single GPU.

Differential Revision: [D34433514](https://our.internmc.facebook.com/intern/diff/D34433514/)

[ghstack-poisoned]

fegin reviewed (approved):
LGTM
